Page titles (and drive-by channel actions refactor)

indirectlylit opened the pull request:
This PR adds unique page titles to every major page-switching action in the Learn and Management apps.

While working on this, I got sucked into a refactor of channels and Promises, mainly in learn's `actions.js`. High-level changes there:

* Made helper functions synchronous instead of returning Promises.
* Now treats the Channels resource more consistently with other resources.
* Added a `_channelListState` mapper function, similar to other state mappers
* Added explicit checks channel ID availability
* Consolidated cookie handling into helper functions

indirectlylit commented:
Note that I've done a fair amount of testing, but there is a true risk of regression introduction here because so much of the logic has changed.

radinamatic commented:
I'll pull this branch to try locally, what do you recommend should be tested more thoroughly?

indirectlylit commented:
Unfortunately, the logic related to _all_ channel switching, browser history, and navigation in the learn app.

I'm also somewhat inclined to wait to merge this until after we send UNICEF the installer. However if we get some eyes on it and it seems pretty stable, the risk might be low enough to merge sooner.

rtibbles commented on kolibri/plugins/learn/assets/src/actions.js:
```diff
@@ -70,74 +72,77 @@ function _collectionState(data) {
+  cookiejs.set('currentChannelId', currentChannelId);
+
+  if (!channelList.find(channel => channel.id === currentChannelId)) {
+    store.dispatch('CORE_SET_ERROR', 'Channel not found');
```
I think this pattern could be abstracted.

rtibbles commented on kolibri/plugins/learn/assets/src/actions.js:
```diff
@@ -70,74 +72,77 @@ function _collectionState(data) {
+ * Set channel state info.
+ * Returns `true` if the requested channel ID is available in the list.
+ */
+function _handleChannels(store, currentChannelId, channelList) {
```
Not entirely comfortable with its combined use as a check, given that it has side effects. Could we have one function that just does the check, and another that is an action that does the mutation dispatch?

rtibbles commented on kolibri/plugins/learn/assets/src/actions.js:
```diff
@@ -151,117 +156,109 @@ function redirectToExploreChannel(store) {
+    (channelsData) => {
+      const channelList = _channelListState(channelsData);
+      const channelId = _getDefaultChannelId(store, channelList);
+      _handleChannels(store, channelId, channelList);
```
For example, this will throw an error because the channel is not found, but then the router will carry on and try to execute the action to set the page state based on this channelId.

rtibbles commented on kolibri/plugins/learn/assets/src/actions.js:
```diff
@@ -151,117 +156,109 @@ function redirectToExploreChannel(store) {
+    (channelsData) => {
+      const channelList = _channelListState(channelsData);
+      const channelId = _getDefaultChannelId(store, channelList);
+      _handleChannels(store, channelId, channelList);
```
As above with explore. Is there a way we could avoid repeating this code between the two, also?

rtibbles commented on kolibri/plugins/learn/assets/src/actions.js:
```diff
@@ -151,117 +156,109 @@ function redirectToExploreChannel(store) {
 
 
-function showExploreTopic(store, channelId, id) {
+function showExploreTopic(store, channelId, id, isRoot = false) {
```
Much DRYer, I like it.

rtibbles commented on kolibri/plugins/learn/assets/src/actions.js:
```diff
@@ -151,117 +156,109 @@ function redirectToExploreChannel(store) {
-      const pageState = { id };
-      pageState.topic = _topicState(attributes);
+    ([topic, children, channelsData]) => {
+      if (!_handleChannels(store, channelId, _channelListState(channelsData))) {
```
This makes sense, because we are presumably just setting the error page here, and giving up.

rtibbles commented on kolibri/plugins/learn/assets/src/actions.js:
```diff
@@ -151,117 +156,109 @@ function redirectToExploreChannel(store) {
       store.dispatch('CORE_SET_ERROR', null);
+      if (isRoot) {
+        const currentChannel = getters.currentChannel(store.state);
+        document.title = _explorePageTitle(currentChannel.title);
```
A general thought - do we want to always prefix the title e.g. `Kolibri | Explore - UNICEF Sports Academy`?

rtibbles replied:
@radinamatic please weigh in here, I imagine such prefixes might be annoying for screenreaders?

radinamatic replied:
Yes, tota11y right @rtibbles, annoying and not informative as it could be. It's also the question of order of prefixes, so IMHO the page title should ideally contain: `H1` + `Channel Name` + `Kolibri`, or `H1` + `Facility Name` + `Kolibri` (for admin management pages). 

Examples:

`<title>Explore - Khan Academy - Kolibri</title>`

![selection_117](https://cloud.githubusercontent.com/assets/1457929/18971687/45c924ae-8696-11e6-99ca-4549c2871a20.png)

**Adding type of content might also be useful**
`<title>Addition and subtraction intro - Topic - Khan Academy - Kolibri</title>`

![selection_118](https://cloud.githubusercontent.com/assets/1457929/18971751/8f2f6efa-8696-11e6-855f-378e0e5dcba3.png)

`<title>Add and subtract pieces of fruit - Video - Khan Academy - Kolibri</title>`

![selection_119](https://cloud.githubusercontent.com/assets/1457929/18971889/1d48d3f2-8697-11e6-95f5-6e417a9a7777.png)

`<title>All Users - Manage - (Facility-Name) - Kolibri</title>`

![selection_120](https://cloud.githubusercontent.com/assets/1457929/18971974/812da820-8697-11e6-8460-9b0f645eae49.png)

rtibbles commented on kolibri/plugins/learn/assets/src/actions.js:
```diff
@@ -151,117 +156,109 @@ function redirectToExploreChannel(store) {
+      }
     },
     (error) => {
       store.dispatch('CORE_SET_ERROR', JSON.stringify(error, null, '\t'));
```
Yeah, I think we can probably abstract this error handler?

```
const errorHandler = (error) => {
    store.dispatch('CORE_SET_ERROR', error instanceof String ? error : JSON.stringify(error, null, '\t'));
    store.dispatch('CORE_SET_PAGE_LOADING', false);
    document.title = _errorTitle();
};
```

rtibbles commented on kolibri/plugins/learn/assets/src/actions.js:
```diff
@@ -151,117 +156,109 @@ function redirectToExploreChannel(store) {
+
+  ChannelResource.getCollection({}).fetch().then(
+    (channelsData) => {
+      if (!_handleChannels(store, channelId, _channelListState(channelsData))) {
```
Should definitely change this to a check, I think? Could then do (or this could be abstracted further):

```
if (!_checkChannels(store, channelId, _channelListState(channelsData))) {
    handleError('Channel not Found');
    return;
}
```

rtibbles commented on kolibri/plugins/learn/assets/src/actions.js:
```diff
@@ -70,74 +72,77 @@ function _collectionState(data) {
+  if (!channelList.find(channel => channel.id === currentChannelId)) {
+    store.dispatch('CORE_SET_ERROR', 'Channel not found');
+    store.dispatch('CORE_SET_PAGE_LOADING', false);
+    document.title = _errorTitle();
```
I have been trying to nail down what bothers me about this, and I think it's the fact that this is not reactive, and so the document title is not represented in state.

I think we should do this by having a core-title component that can be passed in the right title based on Vuex state (this would also allows us to move the title names into Vue components and solve our $trs issue simply).

radinamatic commented:
@indirectlylit Any reason page titles are not visible in Firefox when running this branch locally?

indirectlylit commented:
@radinamatic - not that I'm aware of. I'll check it out when I make updates as per @rtibbles 's review

radinamatic commented:
Weird, in my FF on GNOME Ubuntu, I see no reflection of the title even though it's there when I inspect... :thinking: 

![kolibri dev server - mozilla firefox_126](https://cloud.githubusercontent.com/assets/1457929/19084563/6307728a-8a67-11e6-81d0-2a60fc76ac48.png)

 Wonder if it has something to do with my setup, I'll restart without add-ons (Safe mode) to check!

radinamatic commented:
Mystery solved, all's well when I restart in safe mode! :+1: 

![learn - addition and subtraction within 10 - mozilla firefox_127](https://cloud.githubusercontent.com/assets/1457929/19084868/75788e9e-8a68-11e6-8814-66811daa9e1b.png)

This non-issue aside, my above recommendation stands - page titles should be distinctive, so consider using:

1. `H1` + (`content-type`) +  `Channel Name` + `Kolibri` for user facing pages. 

At this point I'm not sure if `Learn` and `Explore` distinction has much relevance once the user is browsing content, so those prefixes might really be only repetitive and not necessary *except* on `Learn` and `Explore` root pages.  

2. On second thought, your current solution for admin management pages seems perfect! :wink: 

@indirectlylit

indirectlylit commented:
I've addressed the comments in here with one exception: I didn't add the 'content kind' to the title. I'll open an enhancement task for this in trello.
